transports/websocket/src/quicksink.rs

```diff
@@ -0,0 +1,351 @@
+                    } else {
+                        return Poll::Ready(Ok(()));
+                    }
+                }
+                State::Sending => match ready!(this.future.as_mut().as_pin_mut().unwrap().poll(cx))
+                {
+                    Ok(p) => {
+                        this.future.set(None);
+                        *this.param = Some(p);
+                        *this.state = State::Empty
+                    }
+                    Err(e) => {
+                        this.future.set(None);
+                        *this.state = State::Failed;
+                        return Poll::Ready(Err(e));
+                    }
+                },
+                State::Flushing => {
+                    match ready!(this.future.as_mut().as_pin_mut().unwrap().poll(cx)) {
+                        Ok(p) => {
+                            this.future.set(None);
+                            *this.param = Some(p);
+                            *this.state = State::Empty
+                        }
+                        Err(e) => {
+                            this.future.set(None);
+                            *this.state = State::Failed;
+                            return Poll::Ready(Err(e));
+                        }
+                    }
+                }
+                State::Closing => match ready!(this.future.as_mut().as_pin_mut().unwrap().poll(cx))
+                {
+                    Ok(_) => {
+                        this.future.set(None);
+                        *this.state = State::Closed;
+                        return Poll::Ready(Ok(()));
+                    }
+                    Err(e) => {
+                        this.future.set(None);
+                        *this.state = State::Failed;
+                        return Poll::Ready(Err(e));
+                    }
+                },
+                State::Closed => return Poll::Ready(Ok(())),
+                State::Failed => panic!("SinkImpl::poll_closed called after error."),
+            }
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::quicksink::{make_sink, Action};
+    use async_std::{io, task};
+    use futures::{channel::mpsc, prelude::*, stream};
+
+    #[test]
+    fn smoke_test() {
+        task::block_on(async {
+            let sink = make_sink(io::stdout(), |mut stdout, action| async move {
+                match action {
+                    Action::Send(x) => stdout.write_all(x).await?,
+                    Action::Flush => stdout.flush().await?,
+                    Action::Close => stdout.close().await?,
+                }
+                Ok::<_, io::Error>(stdout)
+            });
+
+            let values = vec![Ok(&b"hello\n"[..]), Ok(&b"world\n"[..])];
+            assert!(stream::iter(values).forward(sink).await.is_ok())
+        })
+    }
+
+    #[test]
+    fn replay() {
+        task::block_on(async {
+            #[allow(clippy::disallowed_methods)]
+            let (tx, rx) = mpsc::unbounded();
```

Comment:
Can we just use a bounded channel here?

Reply:
Considering it was in a test alone, this was the simplest way forward. I'll evaluate length of the test's runtime...